### PR TITLE
#1395: Focus repo

### PR DESCRIPTION
**Pull request by psymoon**
### Pull Request checklist

- [ ] This PR includes thorough **tests** or an explanation of why it does not
- [ ] This PR includes a **CHANGELOG entry** or does not need one
- [ ] I have considered adding **QA labels** on the associated issue (not this PR; `qa-ready` or `qa-denied`)

### CODE REVIEW

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 1]
```diff
@@ -87,9 +86,14 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
     // work properly, so we !!
     private val youtubeBackHandler by lazy { YouTubeBackHandler(engineView!!, activity as MainActivity) }
 
+    private lateinit var webRenderViewModel: WebRenderViewModel
+    private lateinit var rootView: View
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Let's make sure we remember to null this out in order to prevent leaking.

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+            val newState = State(
+                focusNode = FocusNode(viewId),
+                focused = true)
+
+            _state.onNextIfNew(newState)
+        }
+    }
+
+    @VisibleForTesting
+    private fun dispatchFocusUpdates(
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: We aren't dispatching here anymore, maybe `createFocusNode`?

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            pocketRepo.feedState) { state, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState ->
+        dispatchFocusUpdates(state.focusNode, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState)
+    }
+
+    val focusUpdate: Observable<State> = _focusUpdate
+            .distinctUntilChanged()
+            .hide()
+
+    override fun onGlobalFocusChanged(oldFocus: View?, newFocus: View?) {
+        fun <T> BehaviorSubject<T>.onNextIfNew(value: T) {
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Most of `onNextIfNew` is the same thing as just putting `distinctUntilChanged` on the `Observable`.  We can add that operator, get rid of most of this, and then just have (`if (newState.focusNode.viewId != NESTED_WEB_VIEW_ID) this.onNext(value)`.

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+            val newState = State(
+                focusNode = FocusNode(viewId),
+                focused = true)
+
+            _state.onNextIfNew(newState)
+        }
+    }
+
+    @VisibleForTesting
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: I think it would be good to add a note up here that the majority of the logic is lifted from existing code.  There are some problems that we might address in the future, but for now we're just consolidating it.

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    private fun dispatchFocusUpdates(
+        focusNode: FocusNode,
+        activeScreen: ActiveScreen,
+        sessionState: SessionRepo.State,
+        pinnedTilesIsEmpty: Boolean,
+        pocketState: PocketVideoRepo.FeedState
+    ): State {
+
+        var newState = _state.value!!
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Can we modify the params to include `newState`, instead of reaching in here?

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+            }
+            ActiveScreen.WEB_RENDER -> Unit
+            ActiveScreen.POCKET -> Unit
+            ActiveScreen.SETTINGS -> Unit
+        }
+
+        return newState
+    }
+
+    private fun updateNavUrlInputFocusTree(
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: `updateNavUrlInputFocusTree` -> `getNavUrlInputFocusNode`
Or something along those lines.  We aren't mutating any existing object, so update seems like a misnomer to me.

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+        val nextFocusDownId = when {
+            pocketState is PocketVideoRepo.FeedState.FetchFailed -> R.id.megaTileTryAgainButton
+            pocketState is PocketVideoRepo.FeedState.Inactive -> {
+                if (pinnedTilesIsEmpty) {
+                    R.id.navUrlInput
+                } else {
+                    R.id.tileContainer
+                }
+            }
+            else -> R.id.pocketVideoMegaTileView
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Could we enumerate the other states here, instead of using else?  That way if we add any in the future, this will need to be updated before it'll compile.

`LoadComplete || Loading || NoAPIKey -> R.id.pocketVideoMegaTileView`

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+    sessionRepo: SessionRepo,
+    pinnedTileRepo: PinnedTileRepo,
+    pocketRepo: PocketVideoRepo
+) : ViewTreeObserver.OnGlobalFocusChangeListener {
+
+    val defaultFocusMap: HashMap<ActiveScreen, Int> = HashMap()
+
+    data class State(
+        val focusNode: FocusNode,
+        val focused: Boolean = true
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Since we're exposing a few different things from this class (nodes, default focus requests, etc), I find the `State` name a little confusing.  What else could we call this?

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+        pocketState: PocketVideoRepo.FeedState
+    ): State {
+
+        assert(lostFocusNode.viewId == R.id.tileContainer ||
+                lostFocusNode.viewId == R.id.megaTileTryAgainButton)
+
+        val viewId = when (pocketState) {
+            PocketVideoRepo.FeedState.FetchFailed -> R.id.megaTileTryAgainButton
+            PocketVideoRepo.FeedState.Inactive -> R.id.navUrlInput
+            else -> R.id.pocketVideoMegaTileView
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Same as above

> Could we enumerate the other states here, instead of using else? That way if we add any in the future, this will need to be updated before it'll compile.
>
> LoadComplete || Loading || NoAPIKey -> R.id.pocketVideoMegaTileView

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt (hunk 6 above)
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Same comment for all the other `update` method names.

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+        }
+    }
+
+    private fun updateDefaultFocusForOverlayWhenTransitioningFromWebRender(
+        sessionState: SessionRepo.State
+    ) {
+        defaultFocusMap[ActiveScreen.NAVIGATION_OVERLAY] = when {
+            sessionState.backEnabled -> R.id.navButtonBack
+            sessionState.forwardEnabled -> R.id.navButtonForward
+            else -> R.id.navButtonReload
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Reload isn't guaranteed to always be enabled when returning from WebRender.  We need to fall back to navUrlInput if the current URL is home.

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+            screenController.currentActiveScreen,
+            sessionRepo.state,
+            pinnedTileRepo.isEmpty,
+            pocketRepo.feedState) { state, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState ->
+        dispatchFocusUpdates(state.focusNode, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState)
+    }
+
+    val focusUpdate: Observable<State> = _focusUpdate
+            .distinctUntilChanged()
+            .hide()
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: `_focusUpdate` and `focusUpdate` could be merged into
```
val focusUpdate = Observables.combineLatest(
            _state,
            screenController.currentActiveScreen,
            sessionRepo.state,
            pinnedTileRepo.isEmpty,
            pocketRepo.feedState) { state, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState ->
        dispatchFocusUpdates(state.focusNode, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState)
    }
    .distinctUntilChanged()
```

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: We don't need `hide` anymore because this isn't a subject (I assume it was at an earlier commit).

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+                    R.id.tileContainer
+                }
+            }
+            else -> R.id.pocketVideoMegaTileView
+        }
+
+        val nextFocusUpId = when {
+            sessionState.backEnabled -> R.id.navButtonBack
+            sessionState.forwardEnabled -> R.id.navButtonForward
+            sessionState.currentUrl != URLs.APP_URL_HOME -> R.id.navButtonReload
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Could we add a TODO here and mention that this is duplicating existing logic in the ToolbarVM, and could fall out of sync?

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt [hunk 13]
```diff
@@ -179,10 +171,11 @@ class NavigationOverlayFragment : Fragment() {
         ToolbarUiController(
             toolbarViewModel,
             ::exitFirefox,
-            { updateFocusableViews() },
             onNavigationEvent
         ).onCreateView(view, viewLifecycleOwner, fragmentManager!!)
 
+        rootView = view
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: Let's make sure we null this out in `onViewDestroyed`.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt [hunk 14]
```diff
@@ -282,16 +295,15 @@ class NavigationOverlayFragment : Fragment() {
...
-        updateFocusableViews()
     }
 
     private fun hideMegaTileError() {
         pocketVideosContainer.visibility = View.VISIBLE
         pocketErrorContainer.visibility = View.GONE
-        updateFocusableViews()
+
+        // View.focusable = INT only available Android API > 26 :(
+        pocketVideoMegaTileView.setFocusable(true)
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: What is `View.focusable = INT`?

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt [hunk 15]
```diff
@@ -334,13 +346,12 @@ class NavigationOverlayFragment : Fragment() {
...
             if (it != null) {
                 tileAdapter.setTiles(it)
-                updateFocusableViews()
             }
         })
 
         adapter = tileAdapter
 
-        layoutManager = HomeTileManager(context, COL_COUNT)
+        layoutManager = GridLayoutManager(context, COL_COUNT)
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: This seems unrelated to the other changes, what is it for?

File: app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileRepo.kt [hunk 16]
```diff
@@ -37,6 +37,8 @@ class PinnedTileRepo(private val applicationContext: Application) {
             BehaviorSubject.create()
     val pinnedTiles: Observable<LinkedHashMap<String, PinnedTile>> = _pinnedTiles.hide()
 
+    val isEmpty: Observable<Boolean> = _pinnedTiles.map { it.size == 0 }.distinctUntilChanged()
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileRepo.kt]: :+1:

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileRepo.kt]: Could we add a line break before `distinctUntilChanged`, though?  It's Rx convention to put one operator on a line, although it's still pretty clear what's happening here.

File: app/src/main/res/layout/activity_main.xml [hunk 17]
```diff
@@ -2,35 +2,36 @@
...
 
     <FrameLayout
-            android:orientation="vertical"
-            android:id="@+id/container_web_render"
-            android:layout_width="match_parent"
-            android:layout_height="match_parent"/>
+        android:orientation="vertical"
+        android:id="@+id/container_web_render"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"/>
```
Comment (Baron-Severin) [inline app/src/main/res/layout/activity_main.xml]: Looks like there were some unintended indentation changes in this file.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt (hunk 1 above)
Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Sure, I'd be surprised if they are not already garbage collected `onDestroy` though (like all the other `lateinit vars`). But to be safe (since we are messing with view references) I will make this nullable and clear it up in `onDestroyView`

File: app/src/main/res/layout/activity_main.xml (hunk 17 above)
Comment (psymoon) [reply app/src/main/res/layout/activity_main.xml]: It was improperly indented before actually, this was intended

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt (hunk 10 above)
Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: `defaultFocusMap` is gone now, but yes good catch

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt (hunk 11 above)
Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Cool

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt (hunk 14 above)
Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: View.focusable takes in Int type only, not boolean (unless I'm mistaken)

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt (hunk 7 above)
Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: Since this is a when {} block instead of when(condition), else is necessary

File: app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt (hunk 8 above)
Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt]: State no longer exposes default focus requests, I could rename it FocusState if you'd like.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt (hunk 15 above)
Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: HomeTileManager has custom focusUpId logic that depends on PocketState.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt (hunk 1 above)
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Unless I'm mistaken, `lateinit vars` don't clean themselves up in `onDestroy`.  Is there some  KTX functionality or something that I'm unaware of?
